File: src/sdk/js/text_codec.js
// UTF-8-only TextEncoder / TextDecoder. Conversion happens host-side
// (__highbeam_utf8_*); these classes are the spec-shaped wrappers.
(() => {
    if (globalThis.TextEncoder !== undefined) return;

    class TextEncoder {
        constructor() {
            this.encoding = "utf-8";
        }

        encode(input = "") {
            return __highbeam_utf8_encode(String(input));
        }
    }

    class TextDecoder {
        constructor(label = "utf-8") {
            const enc = String(label).trim().toLowerCase();
            if (enc !== "utf-8" && enc !== "utf8" && enc !== "unicode-1-1-utf-8") {
                throw new RangeError(`TextDecoder: unsupported encoding ${label} (utf-8 only)`);
            }
            this.encoding = "utf-8";
        }

        decode(input) {
            if (input === undefined) return "";
            // Normalise the BufferSource shapes to the Uint8Array the host
            // hook expects; anything else falls through and the host throws.
            const bytes =
                input instanceof Uint8Array ? input
                : input instanceof ArrayBuffer ? new Uint8Array(input)
                : ArrayBuffer.isView(input) ? new Uint8Array(input.buffer, input.byteOffset, input.byteLength)
                : input;
            return __highbeam_utf8_decode(bytes);
        }
    }

    globalThis.TextEncoder = TextEncoder;
    globalThis.TextDecoder = TextDecoder;
})();

File: src/sdk/text_codec.rs
//! UTF-8-only `TextEncoder` / `TextDecoder` polyfill.
//!
//! Raw `QuickJS` ships neither, but `fs.readCache` hands plugins a
//! `Uint8Array` — without a decoder the bytes are a dead end. Conversion
//! runs host-side (Rust strings are UTF-8 already; invalid sequences
//! degrade to U+FFFD like a non-fatal spec decoder). The JS layer is a
//! thin class wrapper; constructing `TextDecoder` with any label other
//! than UTF-8 throws `RangeError`.

use rquickjs::{Ctx, Error as JsError, Function, TypedArray, Value};

use crate::sdk::errors::throw_named;

const TEXT_CODEC_JS: &str = include_str!("js/text_codec.js");

const ENCODE_GLOBAL: &str = "__highbeam_utf8_encode";
const DECODE_GLOBAL: &str = "__highbeam_utf8_decode";

/// Install `TextEncoder` / `TextDecoder` on the global object. Idempotent.
///
/// # Errors
///
/// Propagates JS errors from constructing the host functions or evaluating
/// the wrapper script.
pub fn install<'js>(ctx: &Ctx<'js>) -> Result<(), JsError> {
    let encode = Function::new(ctx.clone(), |ctx: Ctx<'js>, text: String| {
        TypedArray::new(ctx, text.into_bytes()).map(TypedArray::into_value)
    })?;
    ctx.globals().set(ENCODE_GLOBAL, encode)?;

    let decode = Function::new(ctx.clone(), |ctx: Ctx<'js>, value: Value<'js>| {
        let Ok(bytes) = TypedArray::<u8>::from_value(value) else {
            return Err(throw_named(
                &ctx,
                "TypeError",
                "TextDecoder.decode: expected a BufferSource",
            ));
        };
        let slice: &[u8] = bytes.as_ref();
        Ok(String::from_utf8_lossy(slice).into_owned())
    })?;
    ctx.globals().set(DECODE_GLOBAL, decode)?;

    ctx.eval::<(), _>(TEXT_CODEC_JS)
}
